### ISSUE TITLE
tcmalloc: Attempt to realloc invalid pointer under FreeBSD

### ISSUE THREAD

**Issue by iron-udjin** [yandex/ClickHouse#1257]
Hello,

System: FreeBSD 11-STABLE r323687 with Clang 4.0, clickhouse-1.1.54289, google-perfttools-2.6.1_1

I'm not sure is it problem in clickhousre. Possibly in tcmalloc. When I start clickhouse, it works a few seconds and fall.
In stderr log I see:
```
Should logs to /var/log/clickhouse-server/clickhouse-server.log
Should error logs to /var/log/clickhouse-server/clickhouse-server.err.log
src/tcmalloc.cc:288] Attempt to realloc invalid pointer 0x80ae00000 
```
In clickhouse-server.err.log:
```
2017.09.19 01:29:02.076342 [ 7 ] <Error> BaseDaemon: ########################################
2017.09.19 01:29:02.076530 [ 7 ] <Error> BaseDaemon: (from thread 3) Received signal Abort trap (6).
2017.09.19 01:29:02.076619 [ 7 ] <Error> BaseDaemon: 0. 0x809b0a6aa <_thr_kill+10> at /lib/libc.so.7
```

Could you please point me how can I debug this issue?
I have core dump file. Should I recompille **clickhouse** and **google-perfttools** with **-g** CFLAG and try to backtrace core dump via **gdb**?
Thank you!

**Comment by ludv1x** [comment]
Hi!
Could you show a backtrace in the saved core file (`gdb /usr/bin/clickhouse  /var/lib/clickhouse/cores/core -batch -ex bt`)?

Also, you could recompile ClickHouse with `-DDEBUG_LIBTCMALLOC=1` CMake variable for better future diagnostics.

**Comment by ludv1x** [comment]
BTW I found the same errors in our production servers: `tcmalloc.cc:290] Attempt to realloc invalid pointer 0x3f000000`. But only two mistakes of all time in ~700 servers.

**Comment by iron-udjin** [comment]
I've recompilled clickhouse with -DDEBUG_LIBTCMALLOC=1 opition. A new message has been appeared in stderr log file:
`memory allocation bug: object at 0x80b000000 has never been allocated`
Here is backtrace of core file:
```
# gdb `which clickhouse` /var/tmp/800.clickhouse.core -batch -ex bt
(no debugging symbols found)...Core was generated by `/usr/local/bin/clickhouse --server --daemon --pid-file /var/run/clickhouse/click'.
Program terminated with signal 6, Aborted.
(no debugging symbols found)...
warning: Lowest section in /usr/local/lib/libicudata.so.58 is .hash at 0000000000000120
#0  0x0000000809a2af7c in _umtx_op_err () from /lib/libthr.so.3
```
Should I recompille **icu** with debugging symbols?

**Comment by proller** [comment]
you can try build without tcmalloc: -DENABLE_LIBTCMALLOC=0

**Comment by iron-udjin** [comment]
Done. There are no errors in stderr log.
In clickhouse-server.err.log:
```
2017.09.19 23:25:01.343512 [ 7 ] <Error> BaseDaemon: ########################################
2017.09.19 23:25:01.343707 [ 7 ] <Error> BaseDaemon: (from thread 3) Received signal Segmentation fault (11).
2017.09.19 23:25:01.343724 [ 7 ] <Error> BaseDaemon: Address: NULL pointer.
2017.09.19 23:25:01.343801 [ 7 ] <Error> BaseDaemon: 0. 0x8098f7f52 <_pthread_mutex_init_calloc_cb+6354> at /lib/libc.so.7
```

**Comment by vavrusa** [comment]
We just had a similar issue (not sure it's the same thing though), several hosts segfaulted at the same time. Lots of:

```
<Error> BaseDaemon: (from thread 16) Received signal Segmentation fault (11).
<Error> BaseDaemon: Address: NULL pointer.
<Error> BaseDaemon: 0. /usr/bin/clickhouse-server(tcmalloc::ThreadCache::ReleaseToCentralCache(tcmalloc::ThreadCache::FreeList*, unsigned long, int)+0xe3) [0x2fed613]
<Error> BaseDaemon: 1. /usr/bin/clickhouse-server(tcmalloc::ThreadCache::Scavenge()+0x40) [0x2fed770]
<Error> BaseDaemon: 2. /usr/bin/clickhouse-server(operator delete(void*, std::nothrow_t const&)+0x3b6) [0x37b3bb6]
<Error> BaseDaemon: 3. /usr/bin/clickhouse-server(Allocator<false>::free(void*, unsigned long)+0x65) [0x2a8ee15]
<Error> BaseDaemon: 4. /usr/bin/clickhouse-server(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName> >::~vector()+0x14a) [0x10c6eba]
<Error> BaseDaemon: 5. /usr/bin/clickhouse-server(DB::MergeTreeDataMerger::mergePartsToTemporaryPart(std::vector<std::shared_ptr<DB::MergeTreeDataPart const>, std::allocator<std::shared_ptr<DB::MergeTreeDataPart const> > >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::MergeListEntry&, unsigned long, long, DB::DiskSpaceMonitor::Reservation*, bool)+0x2bb8) [0x2c252c8]
<Error> BaseDaemon: 6. /usr/bin/clickhouse-server(DB::StorageReplicatedMergeTree::executeLogEntry(DB::ReplicatedMergeTreeLogEntry const&)+0xfe6) [0x2bb0286]
<Error> BaseDaemon: 7. /usr/bin/clickhouse-server() [0x2bb2c1d]
<Error> BaseDaemon: 8. /usr/bin/clickhouse-server(DB::ReplicatedMergeTreeQueue::processEntry(std::function<std::shared_ptr<zkutil::ZooKeeper> ()>, std::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&, std::function<bool (std::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>)+0x4a) [0x2c9c25a]
<Error> BaseDaemon: 9. /usr/bin/clickhouse-server(DB::StorageReplicatedMergeTree::queueTask()+0x150) [0x2b9dc50]
<Error> BaseDaemon: 10. /usr/bin/clickhouse-server(DB::BackgroundProcessingPool::threadFunction()+0x3de) [0x2bf9aae]
<Error> BaseDaemon: 11. /usr/bin/clickhouse-server() [0x379e66f]
<Error> BaseDaemon: 12. /lib/x86_64-linux-gnu/libpthread.so.0(+0x7494) [0x7fe95ccc8494]
```

I attached a full log. There's a couple of instances crashing in quantiles state serialisation, which could possibly be large, but it could be just a coincidence.

cc @bobrik 
[tcmalloc_segfault.txt](https://github.com/yandex/ClickHouse/files/1322391/tcmalloc_segfault.txt)

**Comment by felixendres** [comment]
I have never used clickhouse, just found this issue via google, but for what it's worth:
I get
> src/tcmalloc.cc:281] Attempt to realloc invalid pointer 0xbaa900 
> Fatal error 6: Aborted

when starting _emacs_ with LD_PRELOAD=/usr/lib/libtcmalloc_minimal.so.4.1.2
This suggests it's a bug in tcmalloc, however, it also happens with libjemalloc, another memory allocation library. Without LD_PRELOAD being set emacs works fine.

So, yours is probably a bug that is only triggered in combination with tcmalloc, but not necessarily in tcmalloc.

**Comment by proller** [comment]
fixed here fa65bdcf4847fdf7a5a16a6d25f46798432bec41
